// @escape.tech/graphql-armor-max-depth/dist/graphql-armor-max-depth.esm.js
import { GraphQLError, Kind } from 'graphql';

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

const maxDepthDefaultOptions = {
  n: 6,
  ignoreIntrospection: true
};

class MaxDepthVisitor {
  constructor(context, onError, options) {
    _defineProperty(this, "OperationDefinition", void 0);

    _defineProperty(this, "context", void 0);

    _defineProperty(this, "config", void 0);

    _defineProperty(this, "onError", void 0);

    this.context = context;
    this.config = Object.assign({}, maxDepthDefaultOptions, ...Object.entries(options ?? {}).map(([k, v]) => v === undefined ? {} : {
      [k]: v
    }));
    this.onError = onError;
    this.OperationDefinition = {
      enter: this.onOperationDefinitionEnter
    };
  }

  onOperationDefinitionEnter(operation) {
    const depth = this.countDepth(operation);

    if (depth > this.config.n) {
      this.onError('Query is too deep.');
    }
  }

  countDepth(node, parentDepth = 0) {
    var _node$name;

    if (this.config.ignoreIntrospection && 'name' in node && ((_node$name = node.name) === null || _node$name === void 0 ? void 0 : _node$name.value) === '__schema') {
      return 0;
    }

    let depth = parentDepth;

    if ('selectionSet' in node && node.selectionSet) {
      for (let child of node.selectionSet.selections) {
        depth = Math.max(depth, this.countDepth(child, parentDepth + 1));
      }
    }

    if (node.kind == Kind.FRAGMENT_SPREAD) {
      const fragment = this.context.getFragment(node.name.value);

      if (fragment) {
        depth = Math.max(depth, this.countDepth(fragment, parentDepth + 1));
      }
    }

    return depth;
  }

}

const maxDepthRule = (onError, options) => context => new MaxDepthVisitor(context, onError, options);

const maxDepthPlugin = options => {
  return {
    onValidate({
      addValidationRule
    }) {
      addValidationRule(maxDepthRule(msg => {
        throw new GraphQLError(msg);
      }, options));
    }

  };
};

export { maxDepthPlugin, maxDepthRule };
