// @escape.tech/graphql-armor-max-aliases/dist/graphql-armor-max-aliases.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var graphql = require('graphql');

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

const maxAliasesDefaultOptions = {
  n: 15
};

class MaxAliasesVisitor {
  constructor(context, onError, options) {
    _defineProperty(this, "OperationDefinition", void 0);

    _defineProperty(this, "context", void 0);

    _defineProperty(this, "config", void 0);

    _defineProperty(this, "onError", void 0);

    this.context = context;
    this.config = Object.assign({}, maxAliasesDefaultOptions, ...Object.entries(options ?? {}).map(([k, v]) => v === undefined ? {} : {
      [k]: v
    }));
    this.onError = onError;
    this.OperationDefinition = {
      enter: this.onOperationDefinitionEnter
    };
  }

  onOperationDefinitionEnter(operation) {
    const aliases = this.countAliases(operation);

    if (aliases > this.config.n) {
      this.onError('Too many aliases.');
    }
  }

  countAliases(node) {
    let aliases = 0;

    if ('alias' in node && node.alias) {
      ++aliases;
    }

    if ('selectionSet' in node && node.selectionSet) {
      for (let child of node.selectionSet.selections) {
        aliases += this.countAliases(child);
      }
    } else if (node.kind === 'FragmentSpread') {
      const fragment = this.context.getFragment(node.name.value);

      if (fragment) {
        aliases += this.countAliases(fragment);
      }
    }

    return aliases;
  }

}

const maxAliasesRule = (onError, options) => context => new MaxAliasesVisitor(context, onError, options);

const maxAliasesPlugin = options => {
  return {
    onValidate({
      addValidationRule
    }) {
      addValidationRule(maxAliasesRule(msg => {
        throw new graphql.GraphQLError(msg);
      }, options));
    }

  };
};

exports.maxAliasesPlugin = maxAliasesPlugin;
exports.maxAliasesRule = maxAliasesRule;
